src/driver-sinowealth-nubwo.c

```diff
@@ -0,0 +1,315 @@
+};
+
+struct
+sinowealthnubwo_aesthetic_report {
+	uint8_t report_id; //0x03
+	uint8_t cmd[7];
+	uint8_t r;
+	uint8_t g;
+	uint8_t b;
+	uint8_t color_mode;
+	uint8_t padzero; // 0x00
+	uint8_t brightness; // 0x01 to 0x03
+	uint8_t tempo; // 0x05 0x03 0x01
+	uint8_t padzero2[16*3+1];
+} __packed__;
+
+static uint8_t AESTHETIC_CMD[] = {0x06, 0xbb, 0xaa, 0x2a, 0x00, 0x0a, 0x00};
+
+static int
+sinowealthnubwo_test_hidraw(struct ratbag_device *device) 
+{
+	return 
+		ratbag_hidraw_has_report(device, SINOWEALTHNUBWO_AESTHETIC_CMD_REPORTID)
+		&& ratbag_hidraw_has_report(device, SINOWEALTHNUBWO_PERF_CMD_REPORTID)
+		&& ratbag_hidraw_has_report(device, SINOWEALTHNUBWO_GET_FIRMWARE_CMD_REPORTID);
+}
+
+static char*
+sinowealth_get_firmware_string(struct ratbag_device *device)
+{
+	uint8_t buff[SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE];
+	int size;
+
+	size = ratbag_hidraw_set_feature_report(device, SINOWEALTHNUBWO_PERF_CMD_REPORTID, PREFIRMWARE_QUERY_MSG, ARRAY_LENGTH(PREFIRMWARE_QUERY_MSG));
+	if (size < 0)
+		log_error(device->ratbag, "Error while sending pre-firmware request message: %d\n", size);
+
+	size = ratbag_hidraw_get_feature_report(device, SINOWEALTHNUBWO_GET_FIRMWARE_CMD_REPORTID, buff, SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE);
+	if (size != SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE)
+		log_error(device->ratbag ,"Firmware report reply size mismatch expected %d got %d\n", SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE, size);
+	
+	char *fw_str = zalloc(SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE); //Purposefully overallocate zeros to prevent buffer overrun.
+	memcpy(fw_str, buff + SINOWEALTHNUBWO_GET_FIRMWARE_MSGOFFSET, SINOWEALTHNUBWO_GET_FIRMWARE_MSGSIZE - SINOWEALTHNUBWO_GET_FIRMWARE_MSGOFFSET);
+	return fw_str;
+}
+
+static int
+sinowealthnubwo_probe(struct ratbag_device *device)
+{
+	int error;
+	struct ratbag_profile *profile;
+	struct ratbag_resolution *resolution;
+	struct ratbag_led *led;
+	
+	error = ratbag_find_hidraw(device, sinowealthnubwo_test_hidraw);
+	if (error)
+		return error;
+
+	ratbag_device_init_profiles(device,
+			SINOWEALTHNUBWO_NUM_PROFILES,
+			SINOWEALTHNUBWO_NUM_RESOLUTIONS,
+			SINOWEALTHNUBWO_NUM_BUTTONS,
+			SINOWEALTHNUBWO_NUM_LEDS);
+
+	ratbag_device_for_each_profile(device, profile) {
+		profile->is_active = true;
+
+		ratbag_profile_set_cap(profile, RATBAG_PROFILE_CAP_WRITE_ONLY);
+		ratbag_profile_set_report_rate_list(profile, REPORT_RATES, ARRAY_LENGTH(REPORT_RATES));
+		ratbag_profile_for_each_resolution(profile, resolution) {
+			ratbag_resolution_set_dpi_list(resolution, DPILIST, ARRAY_LENGTH(DPILIST));
+			resolution->dpi_x = resolution->dpi_y = DPILIST[0];
+			resolution->is_active = true;
+			resolution->is_default = true;
+		}
+		ratbag_profile_for_each_led(profile, led) {
+			led->mode = RATBAG_LED_OFF;
+			led->color.red = led->color.green = led->color.blue = 0;
+			led->colordepth = RATBAG_LED_COLORDEPTH_RGB_888;
+			led->type = RATBAG_LED_TYPE_SIDE;
+			led->color.red = led->color.green = led->color.blue = 0;
```

Comment:
If I read your code correctly, then the code isn't reporting the current color of the LED back to ratbag, right? Is the sinowealthnubwo_aesthetic_report read-only?

Reply:
That's correct. The official configuration interface can't read the config from the device. (except for dpi)